flox/xarray.py

```diff
@@ -216,76 +222,80 @@ def xarray_reduce(
         raise NotImplementedError
 
     # eventually drop the variables we are grouping by
-    maybe_drop = [b for b in by if isinstance(b, str)]
+    maybe_drop = [b for b in by if isinstance(b, Hashable)]
     unindexed_dims = tuple(
         b
-        for b, isbin_ in zip(by, isbin)
-        if isinstance(b, str) and not isbin_ and b in obj.dims and b not in obj.indexes
+        for b, isbin_ in zip(by, isbins)
+        if isinstance(b, Hashable) and not isbin_ and b in obj.dims and b not in obj.indexes
     )
 
-    by: tuple[DataArray] = tuple(obj[g] if isinstance(g, str) else g for g in by)  # type: ignore
+    by_da = tuple(obj[g] if isinstance(g, Hashable) else g for g in by)
 
     grouper_dims = []
-    for g in by:
+    for g in by_da:
         for d in g.dims:
             if d not in grouper_dims:
                 grouper_dims.append(d)
 
-    if isinstance(obj, xr.DataArray):
-        ds = obj._to_temp_dataset()
-    else:
+    if isinstance(obj, xr.Dataset):
```

Comment:
this rearrangement was weird. Is it a mypy bug?

Reply:
This is the error you get if you isinstance with DataArray:
```python
   # obj: Union[T_Dataset, T_DataArray]
    if isinstance(obj, xr.DataArray):
        ds = obj._to_temp_dataset() # -> xr.Dataset
    else:
        ds = obj  # error: Incompatible types in assignment (expression has type "Union[T_Dataset, T_DataArray]", variable has type "Dataset")
```
My understanding is that mypy always uses the typing from the first time it is defined (`ds: xr.Dataset` narrower typing). It is similar to the typing issues when importing optional modules